sms_analysis.ipynb

```diff
@@ -779,6 +779,149 @@
     "    top_n=widgets.IntSlider(min=15, max=100, step=1, value=5, description='Max words to show:')\n",
     ")"
    ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "from sklearn.cluster import KMeans, MiniBatchKMeans\n",
+    "from sklearn.decomposition import TruncatedSVD\n",
+    "from sklearn.manifold import TSNE\n",
+    "\n",
+    "def _top_words_by_cluster_from_tfidf(\n",
+    "    cluster_id,\n",
+    "    x_data_tfidf_sender,\n",
+    "    tfidf_per_sender_cluster,\n",
+    "    top_n=15,\n",
+    "):\n",
+    "    \"\"\"\n",
+    "    Returns a dataframe of the top words for each cluster by their TFIDF score.\n",
+    "    \n",
+    "    To determine the \"top\", we look at one cluster's TFIDF - avg(other clusters' TFIDFs)\n",
+    "    \n",
+    "    Args:\n",
+    "        cluster_id: The cluster we want to find the top words for\n",
+    "        x_data_tfidf_sender: TFIDF matrix with as many rows as entries in tfidf_per_sender_cluster\n",
+    "        tfidf_per_sender_cluster: Cluster assignment for each entry in x_data_tfidf_sender\n",
+    "        top_n: Number of top words per cluster to include in the result\n",
+    "    \"\"\"\n",
+    "    this_cluster_indicies = np.argwhere(tfidf_per_sender_cluster == cluster_id)\n",
+    "    other_cluster_indicies = np.argwhere(tfidf_per_sender_cluster != cluster_id)\n",
+    "    this_cluster_records = x_data_tfidf_sender[this_cluster_indicies.flatten()]\n",
```

Comment:
Why do you need to flatten anything here?

And can't you do something like:
x_data_tfidf_sender[tfidf_per_sender_cluster == cluster_id] instead of the argwhere and flattering?